### DISCUSSION TITLE
Manually updating parameters - documentation traverse util

### DISCUSSION THREAD

**Discussion by GJBoth** [google/flax#1073] [Q&A] [answered]
I'm working on implementing reservoir computing / echo state networks in Jax & Flax. (We're seeing an almost a 2x speedup compared to pytorch code, so I hope to release it soon!) After training I want to manually set the weights in the output layer. 

Currently I use
```
unfreeze(params)
params['output_layer'] = xxx
freeze(params
```
which works, but I think the preferred method is to use a traverse util? Is this indeed the case, and if so, is there a little bit more documentation on how to apply these? The example in the documentation is sparse and I'm having issues making it work... Thanks!

**Comment by marcvanzee** [comment, marked as answer]
I think you proposed solution works fine and I don't think it is necessary to use `traverse_util` if you simply want to update a subtree of your variables. The `unfreeze`/`freeze` pattern is quite common, and it is for instance also used in our [Model Surgery HOWTO](https://flax.readthedocs.io/en/latest/howtos/model_surgery.html).

`traverse_util` can be usee for more complex operations. For instance, if you would like to replace all occurrences of `output_layer` in your params with a different subtree. For more details on how this can be used, it may be insightful to take a look at some of the tests in [tests/traverse_util_test.py](https://github.com/google/flax/blob/master/tests/traverse_util_test.py).

**Reply by GJBoth** [reply]
In that case, I'll keep it this way. Dankjewel :)

**Comment by Wotaker** [comment]
Hi, how is the progress with your implementation? I have opened a new [thread](https://github.com/google/flax/discussions/3227#discussion-5442418), where I mention this discussion.
